src/spock_apply.c

```diff
@@ -2772,6 +2812,19 @@ apply_work(PGconn *streamConn)
 					if (last_received < endpos)
 						last_received = endpos;
 
+					/*
+					 * It is important to update received_lsn on a keepalive
+					 * message: last_received tells us the last WAL position
+					 * that was processed by the remote walsender, even if that
+					 * data has never be sent to our replica.
+					 * It allows Spock to maintain LSN lag statistic.
+					 *
+					 * NOTE: we can't change the keepalive message format, so
+					 * just apply the same last_inserted. It may cause negative
+					 * delta, but it seems not important.
+					 */
+					UpdateWorkerStats(last_received, last_inserted);
+
 					if (reply_requested)
 						check_and_update_progress(endpos, GetCurrentTimestamp());
```

Comment:
This looks suspicious to me. These functions UpdateWorkerStats and check_and_update_progress write remote_insert_lsn. Two writers on the same field can undo each other, interfering the progress. I suggest to have a single writer, maybe combine them both.

Reply:
Ok, I removed check_and_update_progress